R/display.R

```diff
@@ -1,74 +1,136 @@
+namedlist <- function() setNames(list(), character(0))
+
 base_display <- function(data, metadata) {
     warning("IR_display can only be used from the IPython R kernel and R magic.")
 }
 
 #'Display data by mimetype, with optional alternative representations.
 #'
-#' @param data A named list mapping mimetypes to content (base64 encoded for binary data)
-#' @param metadata A named list mapping mimetypes to named lists of specific metadata
+#' @param data  A named list mapping mimetypes to content (base64 encoded for binary data)
+#' @param metadata  A named list mapping mimetypes to named lists of specific metadata
 #' @export
-display <- function(data, metadata=NULL) {
+display <- function(data, metadata = NULL) {
     base_display(data, metadata)
 }
 
 #'Display a single type of data
 #'
-#' @param mimetype The mimetype of the data
-#' @param  content The data as a length 1 string vector
-#' @param  metadata A named list of metadata
+#' @param mimetype  The mimetype of the data
+#' @param content  The data as a length 1 string vector
+#' @param metadata  A named list of metadata
+#' @export
+display1 <- function(mimetype, content, metadata = NULL) {
+    data <- list()
+    data[[mimetype]] <- content
+    base_display(data, metadata)
+}
+
+#'Display a set of formatted objects that the Jupyter frontend/output can select from
+#'
+#' @param metadata  Initial metadata. Mimetype-wise data will be added
+#' @param ...  Mimetypes and content.
+#' 
+#' You can either give raw vectors or character vectors as content.
+#' 
+#' Also you can give a list of which the first element will be content, the rest metadata.
+#' 
+#' @examples
+#' \dontrun{
+#' display_alternatives(
+#'   'text/html'       = '<em>Wop</em>'
+#'   'application/pdf' = base64encode('test.pdf'),
+#'   'image/png'       = list(some.png.data, width = 800, height = 600),
+#'   'image/svg+xml'   = '<svg></svg>')
+#' }
+#' 
+#' @importFrom base64enc base64encode
 #' @export
-display1 <- function(mimetype, content, metadata=NULL) {
-    data = list()
-    data[mimetype] = content
+display_alternatives <- function(..., metadata = namedlist()) {
+    contents <- list(...)
+    
+    data <- list()
+    for (mime in names(contents)) {
+        content <- contents[[mime]]
+        if (is.list(content)) {
+            metadata[[mime]] <- content[-1]
+            content <- content[[1]]
+        }
+        
+        if (is.raw(content))
```

Comment:
In python, this is done more or less before the 'sending'. IMO it's better to do the base encoding late as that also catches the `display_png/...` cases (see zmqshell.py -> `ZMQDisplayPublisher` and jsonutils.py -> `encode_images()`; sorry can't give a better filenames, I still have the "before the big split" version here :-/)

It basically does this for each image mimetype:

```
    pngdata = format_dict.get('image/png')
    if isinstance(pngdata, bytes):
        # make sure we don't double-encode
        if not pngdata.startswith(PNG64):
            pngdata = encodebytes(pngdata) # base64...
        encoded['image/png'] = pngdata.decode('ascii')
```